ouroboros-network/src/Ouroboros/Network/Protocol/VersionNegotiation/Codec.hs

```diff
@@ -0,0 +1,127 @@
+import           Data.Map (Map)
+import qualified Data.Map as Map
+
+import qualified Codec.CBOR.Encoding as CBOR
+import qualified Codec.CBOR.Decoding as CBOR
+import qualified Codec.CBOR.Read     as CBOR
+import qualified Codec.CBOR.Term     as CBOR
+import           Codec.Serialise (Serialise)
+import qualified Codec.Serialise     as CBOR
+
+import           Network.TypedProtocol.Codec hiding (encode, decode)
+import           Ouroboros.Network.Codec (mkCodecCborLazyBS)
+
+import           Ouroboros.Network.Protocol.VersionNegotiation.Type
+
+
+-- |
+-- Decoding proposed version is done in two stages.  First decode
+-- a @'CBOR.Term'@, this is done by @'codecVersionNegotiation'@, the second
+-- phase is handled by interface defined in this type class and it is used by
+-- @'versionNegotiationClientPeer'@ and @'versionNeogitationServerPeer'@.
+--
+-- todo: find a better place for this class
+class SerialiseTerm a where
+  encodeTerm :: a -> CBOR.Term
+  decodeTerm :: CBOR.Term -> Maybe a
+
+-- |
+-- @'VersionNegotiationProtocol'@ codec.  The @'MsgProposeVersions'@ encodes
+-- proposed map in ascending order and it expects to receive them in this order.
+-- This allows to construct the map in linear time.  There is also another
+-- limiting factor to the number of versions on can present: the whole message
+-- must fit into a single TCP segment.
+--
+codecVersionNegotiation
+  :: forall vNumber m.
+     ( Monad m
+     , MonadST m
+     , Ord vNumber
+     , Enum vNumber
+     , Serialise vNumber
+     , Show vNumber
+     )
+  => Codec (VersionNegotiationProtocol vNumber CBOR.Term) CBOR.DeserialiseFailure m ByteString
+codecVersionNegotiation = mkCodecCborLazyBS encode decode
+    where
+      encode :: forall (pr :: PeerRole) st st'.
+                PeerHasAgency pr st
+             -> Message (VersionNegotiationProtocol vNumber CBOR.Term) st st'
+             -> CBOR.Encoding
+
+      encode (ClientAgency TokPropose) (MsgProposeVersions vs) =
+        let vs' = Map.toAscList vs
+        in
+           CBOR.encodeListLen 2
+        <> CBOR.encodeWord 0
+        <> CBOR.encodeMapLen (fromIntegral $ length vs')
+        <> mconcat [    CBOR.encode vNumber
+                     <> CBOR.encodeTerm vBlob 
+                   | (vNumber, vBlob) <- vs'
+                   ]
+
+      encode (ServerAgency TokConfirm) (MsgAcceptVersion vNumber vBlob) =
+           CBOR.encodeListLen 3
+        <> CBOR.encodeWord 1
+        <> CBOR.encode vNumber
+        <> CBOR.encodeTerm vBlob
+
+      encode (ServerAgency TokConfirm) MsgRefuse =
+           CBOR.encodeListLen 1
+        <> CBOR.encodeWord 2
+
+      encode (ServerAgency TokConfirm) MsgUnknownVersion =
+           CBOR.encodeListLen 1
+        <> CBOR.encodeWord 3
+
+      -- decode a map checking the assumption that
+      --  * keys are different
+      --  * keys are encoded in ascending order
+      -- fail when one of these assumptions is not met
```

Comment:
Is it important to enforce this? It's not immediately clear to me. If it's important lets document why.

Reply:
This puts the requirement of sorting the versions on the requester, this might not be an issue as we also have a constraint on size, but still I'd argue for it.  It is documented in the `MsgProposeVersions`.